### ISSUE TITLE
Reduce false positives on invariant containers initialized with a singleton type (e.g. None)

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#2972]
E.g. `mylist = [None]` or `mydict = {k: None for k in keys}`, etc. This kind of pattern occurs not-infrequently, where the container is later updated with non-None values. It seems like a good guess that the user probably doesn't intend for the container to only ever hold `None`.

The tricky question is what we should widen it to. We could use `Unknown | None`, but it feels a bit weird to introduce a gradual type just in this case.

The best answer might be #1473, where if we see only `None` initially, we look ahead to later additions to decide what wider type to use.

We could also apply this to other singleton types (ellipsis? literals are already handled in promotion), but those cases seem likely so rare as to not really be worth caring about.

**Comment by Jeremiah-England** [comment]
Just updated to the latest `ty` and I think this caused the majority of the new errors.

Most common case (~dozen):

```python
y: list[T | None] = [None] * len(x)
```

Since the codebase is checked with Pyright strict-mode, we already had declared the type of these containers upfront, but `ty` doesn't use it. The fix for the line above is something like:

```python
y: list[T | None] = cast(list[T | None], [None] * len(x))
```

There were two other individual cases I ran into due to the same underlying problem:

(1), function return:

```python
def func() -> list[T | None]:
    if odd_case:
        return [None] * len(x)  # I return `cast(list[T | None], ...)` now.
```

(2) function argument:

```python
def func(x: list[T | None], ...):  # I was able to update to `Sequence[T | None]` to avoid the `cast` below...
    ...

func([None] * len(x), ...)  # Would have had to `cast` here, if `Sequence` wasn't an option.
```

**Comment by ibraheemdev** [comment]
@Jeremiah-England Thanks for the report. Your case is slightly different, and we should be propagating the type context through binary operator dunder calls. Would you mind opening a separate issue with your reproduction?

**Comment by carljm** [comment]
#3013 brings up another kind of singleton type that we might want to promote: class literal objects, e.g. in `[float]`. In this case (unlike with `None` or ellipsis) there are some reasonable promotion candidates more precise than just "union with Unknown". But unlike with int or string literals, there are multiple reasonable candidates (`type[float]`, `type`) and it's not clear which should be picked.

All other type checkers (in the absence of type context) infer `[float]` as `list[type[float]]`. I doubt there's any promotion happening here; I think other type checkers just don't have class literal types in the first place. So I think we should promote to `type[float]`. But that won't help trying to assign to `list[type]` -- for that we'd still need type context.
